support-workers/src/typescript/test/salesforce.it.test.ts

```diff
@@ -0,0 +1,166 @@
+		salesforceService = new SalesforceService(config);
+	});
+
+	test('should be able to upsert a customer', async () => {
+		const result = await salesforceService.upsert(customer);
+		expect(result).toEqual({
+			Success: true,
+			ContactRecord: {
+				Id: salesforceId,
+				AccountId: salesforceAccountId,
+			},
+		});
+	});
+
+	test('should be able to upsert a customer that has optional fields', async () => {
+		const result = await salesforceService.upsert({
+			...customer,
+			OtherStreet: street,
+			OtherCity: city,
+			OtherPostalCode: postCode,
+			OtherCountry: uk,
+			MailingStreet: street,
+			MailingCity: city,
+			MailingPostalCode: postCode,
+			Phone: telephoneNumber,
+		});
+		expect(result).toEqual({
+			Success: true,
+			ContactRecord: {
+				Id: salesforceId,
+				AccountId: salesforceAccountId,
+			},
+		});
+	});
+
+	test('it should be able to add a related contact record', async () => {
+		const name = 'integration-test-recipient';
+		const upsertData: DeliveryContactRecordRequest = {
+			AccountId: salesforceAccountId,
+			Email: emailAddress,
+			Salutation: title,
+			FirstName: name,
+			LastName: name,
+			MailingStreet: street,
+			MailingCity: city,
+			MailingState: state,
+			MailingPostalCode: postCode,
+			MailingCountry: uk,
+			RecordTypeId: '01220000000VB50AAG',
+		};
+		const result = await salesforceService.upsert(upsertData);
+		expect(result).toEqual({
+			Success: true,
+			ContactRecord: {
+				Id: expect.stringMatching('[0-9A-Z]+') as string,
+				AccountId: salesforceAccountId,
+			},
+		});
+	});
+});
+
+jest.mock('../model/stage', () => ({
+	stageFromEnvironment: jest.fn().mockReturnValue('CODE'),
+}));
+
+describe('CreateSalesforceContatctLambda', () => {
+	test('CreateSalesforceContact lambda should upsert a SalesforceContactRecord', async () => {
+		const inputState: CreateSalesforceContactState =
+			createSalesforceContactStateSchema.parse(
+				createSalesforceContactContribution,
+			);
+		const result = await handler(
+			wrapState(inputState, null, {
+				testUser: false,
+				failed: false,
+				messages: [],
+				accountExists: false,
+			}),
+		);
+
+		expect(result.state.product.productType).toBe('Contribution');
+		if (result.state.productSpecificState.productType === 'Contribution') {
+			expect(result.state.productSpecificState.salesForceContact.Id).toBe(
+				'003UD00000bt0YfYAI',
+			);
+			expect(
+				result.state.productSpecificState.salesForceContact.AccountId,
+			).toBe('001UD00000KErfYYAT');
+		}
+	});
+
+	test('should upsert a gift SalesforceContactRecord', async () => {
+		const inputState: WrappedState<CreateSalesforceContactState> =
+			wrapperSchemaForState(createSalesforceContactStateSchema).parse(
+				createGiftSubscription,
+			);
+		const result = await handler(inputState);
+
+		expect(result.state.product.productType).toBe('GuardianWeekly');
+		if (result.state.productSpecificState.productType === 'GuardianWeekly') {
```

Comment:
When would this not be true? Can we control this in the tests? Or is this necessary to keep TypeScript happy somehow?

Reply:
Yes it's just to keep Typescript happy, it would be nice if it could infer it from the expect statement but it can't

Reply:
👍 Would it be worth adding a comment mentioning this?